Create Doctor and Patient class

leezhanpeng opened the pull request:
As MediSync aims to map relationships between people
with different roles, it becomes necessary to distinguish
them beyond their attributes.

The Doctor and the Patient class serves to assist us with
this differentiation.

leezhanpeng committed:
Create Doctor and Patient class
As MediSync aims to map relationships between people
with different roles, it becomes necessary to distinguish
them beyond their attributes.

The Doctor and the Patient class serves to assist us with
this differentiation.

leezhanpeng committed:
Add testing for Doctor and Patient class

leezhanpeng committed:
Add newline at EOF

leezhanpeng committed:
Merge branch 'master' into create-doctor-patient

leezhanpeng committed:
Merge latest codebase

shuenj commented on src/main/java/seedu/address/model/person/Doctor.java:
```diff
@@ -0,0 +1,31 @@
+    /**
+     * Every field must be present and not null.
+     */
+    public Doctor(Name name, Phone phone, Email email, Address address, Set<Affiliation> affiliations) {
```
Perhaps we can update address to role as per latest PR?

leezhanpeng replied:
Yup will do so!

shuenj commented on src/main/java/seedu/address/model/person/Patient.java:
```diff
@@ -0,0 +1,31 @@
+     * Every field must be present and not null.
+     */
+    public Patient(Name name, Phone phone, Email email, Address address, Set<Affiliation> affiliations) {
+        super(name, phone, email, address, affiliations);
```
And perhaps we can have a constructor for Patient/Doctor class that do not require setting the role if it is clearer; i.e. just super(name, phone, email, "Patient", affiliations)

leezhanpeng replied:
Will do that as well 👌

shuenj reviewed (commented):
Just some comments about updating address to role at this stage, and suggestions to distinguish class constructors (not priority imo), otherwise well done!

leezhanpeng committed:
Merge branch 'change-address-to-role' into create-doctor-patient

leezhanpeng committed:
Merge branch 'master' into create-doctor-patient

leezhanpeng committed:
Adapt Doctor and Patient classes to parsers
The current implementation of add and edit command
parsers use the generic Person class.

This reduces the flexibility of what we can do with
different roles as we proceed to implement different
functionality for them.

Let's adapt the newly created Doctor and Patient
classes to the parsers.

leezhanpeng commented on src/main/java/seedu/address/model/person/Role.java:
```diff
@@ -49,6 +51,24 @@ public static boolean isValidRole(String test) {
         return test.matches(VALIDATION_REGEX);
     }
 
+    /**
+     * Returns person with its particular type based on the role.
+     *
+     * @param name Name of person.
+     * @param phone Phone of person.
+     * @param email Email of person.
+     * @param affiliationList Affiliations of person.
+     */
+    public Person generatePerson(Name name, Phone phone, Email email, Set<Affiliation> affiliationList) {
+        if (value.toUpperCase().equals(Type.DOCTOR.name())) {
+            return new Doctor(name, phone, email, affiliationList);
+        } else if (value.toUpperCase().equals(Type.PATIENT.name())) {
+            return new Patient(name, phone, email, affiliationList);
+        } else {
+            return null;
```
As we proceed to use different Person classes, it would be ideal to have the Role class itself generate the correct Person class type. This is in comparison to iterating through the different Person roles and then initialise the correct class during parsing.

However, as the role may intuitively be an attribute of the person, having the Role class generate the Person may not be logical. Open to other possible implementations!

leezhanpeng commented on src/main/java/seedu/address/model/person/Role.java:
```diff
@@ -49,6 +51,24 @@ public static boolean isValidRole(String test) {
         return test.matches(VALIDATION_REGEX);
     }
 
+    /**
+     * Returns person with its particular type based on the role.
+     *
+     * @param name Name of person.
+     * @param phone Phone of person.
+     * @param email Email of person.
+     * @param affiliationList Affiliations of person.
+     */
+    public Person generatePerson(Name name, Phone phone, Email email, Set<Affiliation> affiliationList) {
+        if (value.toUpperCase().equals(Type.DOCTOR.name())) {
+            return new Doctor(name, phone, email, affiliationList);
+        } else if (value.toUpperCase().equals(Type.PATIENT.name())) {
+            return new Patient(name, phone, email, affiliationList);
+        } else {
+            return null;
+        }
+    }
```
As we proceed to use different Person classes, it would be ideal to have the Role class itself generate the correct Person class type. This is in comparison to iterating through the different Person roles and then initialise the correct class during parsing.

leezhanpeng commented on src/main/java/seedu/address/model/person/Role.java:
```diff
@@ -49,6 +51,24 @@ public static boolean isValidRole(String test) {
         return test.matches(VALIDATION_REGEX);
     }
 
+    /**
+     * Returns person with its particular type based on the role.
+     *
+     * @param name Name of person.
+     * @param phone Phone of person.
+     * @param email Email of person.
+     * @param affiliationList Affiliations of person.
+     */
+    public Person generatePerson(Name name, Phone phone, Email email, Set<Affiliation> affiliationList) {
+        if (value.toUpperCase().equals(Type.DOCTOR.name())) {
+            return new Doctor(name, phone, email, affiliationList);
+        } else if (value.toUpperCase().equals(Type.PATIENT.name())) {
+            return new Patient(name, phone, email, affiliationList);
+        } else {
+            return null;
+        }
+    }
```
As we proceed to use different Person classes, it would be ideal to have the Role class itself generate the correct Person class type. This is in comparison to iterating through the different Person roles and then initialise the correct class during parsing.

However, as the role may intuitively be an attribute of the person, having the Role class generate the Person may not be logical. Open to other possible implementations!

shuenj replied:
Yup, I think that is helpful!

shuenj reviewed (approved):
LGTM
